### ISSUE TITLE
Eth1provider logging is broken

### ISSUE THREAD

**Issue by nflaig** [ChainSafe/lodestar#6469]
### Describe the bug

It has been reported by multiple users ([1](https://discord.com/channels/593655374469660673/593655641445367808/1208096150557556779), [2](https://discord.com/channels/593655374469660673/593655641445367808/1210095818925809724)) that in some cases we spam the logs with
```
Eth1Provider is back online oldState=ERROR, newState=ONLINE
```

but the actual error is not logged.

The spam happens because some calls work and some fail, it's likely from the eth1 deposit data tracker.

See related comments
- [updating cache error has been removed](https://github.com/ChainSafe/lodestar/pull/5949/files#r1493363167)
- [error log is never called if errors happen frequently](https://github.com/ChainSafe/lodestar/pull/5949/files#r1493364765)

### Expected behavior

- [ ] We should not spam the logs obviously but we also need to make sure that we properly log errors. It does not make sense to change the state because of a single failed request, good state should be ONLINE or OFFLINE but error it not that helpful, or if we track that state, we need to think about how it is updated as a single successful request changes it to ONLINE again and so on, causing the logs to be spammed with "back online" log above.
- [x] log errors (once a minute) even if errors happen more frequently to provide feedback to user that there is a persistent issue https://github.com/ChainSafe/lodestar/pull/6863

### Steps to reproduce

Based on reports, these two setups can be used to reproduce the issue
- run Lodestar with a syncing execution client, preferably Nethermind as issue has been observed there
- or run [openexecution](https://github.com/tennisbowling/openexecution) as a middleware to connect to EL client

### Additional context

- https://github.com/ChainSafe/lodestar/issues/6128
- https://github.com/ChainSafe/lodestar/pull/5949

### Operating system

Linux

### Lodestar version or commit hash

v1.16.0

**Comment by philknows** [comment]
Can confirm that this is still happening with Lodestar v1.18.0 and Nethermind v1.26.0. 

~~In addition, the logs also include the timeout request errors such as this one:~~
```sh
May-13 10:39:32.918[eth1]            error: Eth1Provider faced error state=ERROR, lastErrorAt=5:44:59 AM - Timeout request
Error: Timeout request
    at JsonRpcHttpClient.fetchJsonOneUrl (file:///usr/app/packages/beacon-node/src/eth1/provider/jsonRpcHttpClient.ts:300:17)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at runNextTicks (node:internal/process/task_queues:64:3)
    at listOnTimeout (node:internal/timers:540:9)
    at processTimers (node:internal/timers:514:7)
    at JsonRpcHttpClient.fetchJson (file:///usr/app/packages/beacon-node/src/eth1/provider/jsonRpcHttpClient.ts:227:16)
    at wrapWithEvents.payload.payload (file:///usr/app/packages/beacon-node/src/eth1/provider/jsonRpcHttpClient.ts:146:37)
    at JsonRpcHttpClient.wrapWithEvents (file:///usr/app/packages/beacon-node/src/eth1/provider/jsonRpcHttpClient.ts:206:24)
    at Eth1Provider.getBlockNumber (file:///usr/app/packages/beacon-node/src/eth1/provider/eth1Provider.ts:186:28)
    at Eth1DepositDataTracker.update (file:///usr/app/packages/beacon-node/src/eth1/eth1DepositDataTracker.ts:197:32)
```

**Comment by nflaig** [comment]
> In addition, the logs also include the timeout request errors such as this one:

This is a different issue, see https://github.com/ChainSafe/lodestar/issues/5935. The log message changed since we refactored the Eth1provider code, but the error is not really related to what is reported in this issue.

**Comment by nazarhussain** [comment]
The only workflow which can trigger such spam is that eth1 providers identified as `providerUrls` return `error/response` alternatively. a.k.a. one request return valid response and next request return error response. This way you will see the `Eth1Provider is back online` very often than the error case, because we log the error once per minute.

**Comment by nflaig** [comment]
> a.k.a. one request return valid response and next request return error response

yeah that's the issue, we need to change the state change logic to not spam the logs, this scenario happens if nethermind is still syncing as it can serve only a subset of requests, doesn't mean it's offline though

> because we log the error once per minute.

if we have continuous errors (frequency < 1min) then we will never log the error again due to a bug in `isOneMinutePassed`.


I guess these are two separate issues and could be fixed independently.

**Comment by nazarhussain** [comment]
Let's consolidate discussion here in this issue, otherwise it will lost on the merge PRs. 

I went over the discussions I had for it's implementation, that was the inspiration or this behavior: 

> also maybe every minute log the error message as well if the eth1 is still down ... ( i.e. if we make the request and its still error and the last error display message > 60 seconds , log the message and update last error display time) 

So yes there is a bug in the implementation, but not in the `createElapsedTimeTracker`. That one seems to be good as what is intended for, it's just tracking time since first invocation. 

For `eth1` provider error we should do via rate limiting helpers for rate limit error per minute.

**Comment by nflaig** [comment]
> So yes there is a bug in the implementation, but not in the createElapsedTimeTracker. That one seems to be good as what is intended for, it's just tracking time since first invocation.

See the test case [here](https://github.com/ChainSafe/lodestar/compare/unstable...nflaig/eth1-error-logging#diff-a01780f44e9510541e78de179880b0a7189c42a6cf18e3d873ca6bc6c2f60e39R55-R66)

I would expect `isOneMinutePassed` to be a rate limiter to log an error max once a minute, but current behavior is to log only once if there are continuous errors with a frequency < 1min. It only will log an error again if there was no error for longer than 1min.

**Comment by nazarhussain** [comment]
The behavior or the `createElapsedTimeTracker` is fine, there is no reset behavior for it. It just track time since it's first creation. 

> Create a tracker which keeps track of the last time a function was called

So if the error appears in first 1 minute of creation of the provider it will not be logged. So we need to change this and instead of using `createElapsedTimeTracker` use some frequency tracker here in the provider.

**Comment by nflaig** [comment]
> So if the error appears in first 1 minute of creation of the provider it will not be logged

this is not true, there is even a test case for it
https://github.com/ChainSafe/lodestar/blob/49c16890a97cc44ea280acfae6712913e5091879/packages/utils/test/unit/waitFor.test.ts#L40

I am pretty sure it will trigger the error log if it's called 1 minute after instantiation


The problem is really that intermediate calls reset the timer, which does not make sense to me. Because the logic of `isOneMinutePassed` should not be "has one minute passed since I was last invoked" but rather "has one minute passed since I last returned `true`".

**Comment by wemeetagain** [comment]
@nflaig can this issue be closed?

**Comment by nflaig** [comment]
the log spam is still a thing
```
Eth1Provider is back online oldState=ERROR, newState=ONLINE
```
although it only happens if you sync the execution client and have only observed it with Nethermind and this should be resolved once we have deposit snapshot as we won't need to query deposit logs from EL

**Comment by nflaig** [comment]
@nazarhussain since this bug has been there for a while, we might even wanna reconsider tracking the status of eth1 provider, it's not really necessary to do this as in 99.9% of cases, this is the same EL as is used for engine api. The other related issue https://github.com/ChainSafe/lodestar/issues/6128 would be solved as well.


It's not clear to me what https://github.com/ChainSafe/lodestar/pull/5949 solved and it just added more UX issues than it solved in my opinion. Not sure if that's easily possible but we should consider reverting that PR.

**Comment by nazarhussain** [comment]
@nflaig If you feel that logging status for the ETH1 provider is not useful we can remove it. 

Or we can change the implementation to only observe periodically not reactively.

**Comment by nflaig** [comment]
this is no longer relevant since ETH1 provider is disabled since electra
